test/test_dataflow.py

```diff
@@ -66,4 +88,147 @@ def test_should_run_beam_job(self):
         # and labels the job
         self.assertEqual(
             driver.pipeline._options.get_all_options()['labels'],
-            ['workflow_id=count_words'])
\ No newline at end of file
+            ['workflow_id=count_words'])
+
+    @patch.object(RunnerResult, 'is_in_terminal_state', create=True)
+    @patch.object(RunnerResult, 'cancel')
+    @patch.object(RunnerResult, 'wait_until_finish')
+    def test_should_run_beam_job_with_timeout_with_cancel(self, wait_until_finish_mock, cancel_mock, is_in_terminal_state_mock):
+        # given
+        wait_until_finish_mock.return_value = 'DONE'
+        is_in_terminal_state_mock.return_value = False
+        driver = CountWordsDriver(Saver(Save()))
+        job = BeamJob(
+            id='count_words',
+            entry_point=driver.run,
+            entry_point_arguments={
+                'words_to_filter': ['valid', 'word'],
+                'words_to_count': ['trash', 'valid', 'word', 'valid']
+            },
+            test_pipeline=self._test_pipeline_with_label('count_words'),
+            execution_timeout=3)
+
+        count_words = Workflow(
+            workflow_id='count_words',
+            definition=[job])
+
+        # when
+        count_words.run('2020-01-01')
+
+        # then
+        self.assertEqual(cancel_mock.call_count, 1)
+        wait_until_finish_mock.assert_called_with(3)
+
+    @patch.object(RunnerResult, 'is_in_terminal_state', create=True)
+    @patch.object(RunnerResult, 'cancel')
+    @patch.object(RunnerResult, 'wait_until_finish')
+    def test_should_run_beam_job_with_timeout_without_cancel(self, wait_until_finish_mock, cancel_mock, is_in_terminal_state_mock):
+        # given
+        wait_until_finish_mock.return_value = 'DONE'
+        is_in_terminal_state_mock.return_value = True
+        driver = CountWordsDriver(Saver(Save()))
+        job = BeamJob(
+            id='count_words',
+            entry_point=driver.run,
+            entry_point_arguments={
+                'words_to_filter': ['valid', 'word'],
+                'words_to_count': ['trash', 'valid', 'word', 'valid']
+            },
+            test_pipeline=self._test_pipeline_with_label('count_words'))
+
+        count_words = Workflow(
+            workflow_id='count_words',
+            definition=[job])
+
+        # when
+        count_words.run('2020-01-01')
+
+        # then
+        self.assertEqual(cancel_mock.call_count, 0)
+        wait_until_finish_mock.assert_called_with(3600000)
+
+    @patch.object(RunnerResult, 'is_in_terminal_state', create=True)
+    @patch.object(RunnerResult, 'cancel')
+    def test_should_run_beam_job_without_timeout_if_wait_until_finish_disabled(self, cancel_mock, is_in_terminal_state_mock):
```

Comment:
You can just validate is that situation occurred and throw the ValueError (if you disabled wait_until_finish, then you can't use the timeout).

Reply:
Changed